core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1413,55 +1412,108 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
-      .pull
-      .timed { timedPull =>
-        def resize(c: Chunk[O], s: Pull[F2, Chunk[O], Unit]): (Pull[F2, Chunk[O], Unit], Chunk[O]) =
-          if (c.size < n) s -> c
-          else {
-            val (unit, rest) = c.splitAt(n)
-            resize(rest, s >> Pull.output1(unit))
-          }
+    fs2.Stream.eval(Semaphore[F2](n.toLong)).flatMap { demand =>
+      fs2.Stream.eval(Semaphore[F2](0L)).flatMap { supply =>
+        fs2.Stream
+          .eval(
+            Ref[F2].of(JunctionBuffer[O](Vector.empty[O], endOfSupply = None, endOfDemand = None))
+          )
+          .flatMap { buffer =>
+            def enqueue(t: O): F2[Boolean] =
+              demand.acquire.flatMap { _ =>
+                buffer
+                  .modify { buf =>
+                    (buf.copy(buf.data :+ t), buf)
+                  }
+                  .flatMap { buf =>
+                    supply.release.map { _ =>
+                      buf.endOfDemand.isEmpty
+                    }
+                  }
+              }
 
-        // Invariants:
-        // acc.size < n, always
-        // hasTimedOut == true iff a timeout has been received, and acc.isEmpty
-        def go(acc: Chunk[O], timedPull: Pull.Timed[F2, O], hasTimedOut: Boolean = false)
-            : Pull[F2, Chunk[O], Unit] =
-          timedPull.uncons.flatMap {
-            case None =>
-              Pull.output1(acc).whenA(acc.nonEmpty)
-            case Some((e, next)) =>
-              def resetTimerAndGo(q: Chunk[O]) =
-                timedPull.timeout(timeout) >> go(q, next)
-
-              e match {
+            def acquireSupplyUpToNWithin(n: Long, timeout: FiniteDuration): F2[Long] =
+              F.race(
+                F.sleep(timeout),
+                supply.acquireN(n)
+              ).flatMap {
                 case Left(_) =>
-                  if (acc.nonEmpty)
-                    Pull.output1(acc) >> resetTimerAndGo(Chunk.empty)
-                  else
-                    go(Chunk.empty, next, hasTimedOut = true)
-                case Right(c) if hasTimedOut =>
-                  // it has timed out without reset, so acc is empty
-                  val (toEmit, rest) =
-                    if (c.size < n) Pull.output1(c) -> Chunk.empty
-                    else resize(c, Pull.done)
-                  toEmit >> resetTimerAndGo(rest)
-                case Right(c) =>
-                  val newAcc = acc ++ c
-                  if (newAcc.size < n)
-                    go(newAcc, next)
-                  else {
-                    val (toEmit, rest) = resize(newAcc, Pull.done)
-                    toEmit >> resetTimerAndGo(rest)
+                  supply.acquire.flatMap { _ =>
+                    supply.available.flatMap { m =>
+                      val k = m.min(n - 1)
+                      supply.tryAcquireN(k).map {
+                        case true  => k + 1
+                        case false => 1
+                      }
+                    }
```

Comment:
Perhaps a for comprehension here?
```suggestion
                  for {
                    _ <- supply.acquire
                    m <- supply.available
                    k  = m.min(n - 1)
                    b <- supply.tryAcquireN(k)
                  } yield if (b) then k + 1 else 1
```

Reply:
❓ Does the default desugaring give worse peformance? I'm thinking of the way that it passes down tuples rather than closing over an outer scoped val. I don't know how much, if any impact that makes once the JIT does its stuff

Reply:
@Daenyth, I'm trying to find `better-monadic-for` string in fs2 sources. However, there is none. Do you know the reason why it hasn't been added?

Reply:
AFAIK it's a general sentiment that it isn't worth the benefit vs deviating from the standard compile. Also it won't work for scala3

Reply:
There is no noticeable performance impact in my test. So I changed the code based on the suggestions, since it looks prettier indeed.